tech_reports/LLMs/llms.md

```diff
@@ -57,12 +57,189 @@ Authors:
+
+<img src="images/row_parallel.png" style="width:500px;"/>
+
+##### **1D Column parallel followed by row parallel (1D weight sharding) **
+
+**When to use:** 1D cluster topologies
+
+1D Weight Sharding is a sharding scheme that combines column and row parallel matmuls and can reduce the data volume sent over CCL operation and thus speed up computation. It consists of a column parallel matmul followed by a row parallel matmul. In this scheme the initial activations are gathered, and the column parallel matmul produces width-sharded outputs. The row parallel matmul consumes those sharded activations and produces parial outputs. We need an AllReduce (ReduceScatter+AllGather) operation to compute the final reduced and gathered outputs.
+
+Optimization potential in this scheme depends highly on the input dimensions to the CCL operations. We can use this scheme for the MLP and any sequence of matmuls that expands and then narrows the output dimension again, becuase it moves the CCL operation to a more beneficial location in the computational graph and thus reduces the CCL data volume.
+
+Let's look at the MLP as concrete example: in Llama3-70b we have `FF1` and `FF3` with dimensions `[32, 8k] x [8k, 28k]` and then the `FF2` with dimension `[32, 28k] x [28k, 8k]`.
+If we gather after `FF1` and `FF3` we have to gather activations of size `[32, 28k/num_devices] -> [32, 28k]` for each of `FF1` and `FF3`; after the `FF2` we'd need to gather again `[32, 8k/num_devices] -> [32, 8k]`.
+If instead, we use the 1D weight sharding scheme and thus move the CCL operation after the `FF2`, we only have to ReduceScatter #num_devices partials of size `[32, 8k] -> [32, 8k/num_devices]` and then optionally AllGather to obtain the `[32, 8k]` gathered outputs.
+
+<img src="images/column_parallel_then_row_parallel.png" style="width:700px;"/>
+
+##### **2D Weight Sharding**
+**When to use:** 2D cluster topologies
+
+In 2D Weight Sharding on a 2D cluster, weights are sharded both in width and height, such that each device contains a block of the weights.
+For this scheme the activations are width-sharded along `cluster_axis=0` and are replicated along `cluster_axis=1`, and the weights are block-sharded. Thus, each device processes a width-shard of the activation, and a block of the weights where the activations are replicated over one axis but the weights are not.
+The matmul result will be width-sharded along `cluster_axis=0` and contain partial results along `cluster_axis=1`.
+Typically an AllReduce (ReduceScatter+AllGather) is used to first reduce along `cluster_axis=1` and then gather the shards along `cluster_axis=0`.
+
+<img src="images/block_sharded.png" style="width:1000px;"/>
+
+##### **Examplary scheme: Llama3**
+
+For our Llama3 family of models we are using the following sharding schemes in our multi-device architectures:
+
```

Comment:
Probably a good idea to link the llama3 attention code file here. Maybe MLP as well. Either in the sentence as a way for the reader to further explore, or link them through the table below. What do you think?

Reply:
Good idea, linked them.